### ISSUE TITLE
Complex rule doesn't get restored correctly

### ISSUE THREAD

**Issue by Betterbird** [Betterbird/thunderbird-patches#256]
Start with
![image](https://github.com/Betterbird/thunderbird-patches/assets/86832723/d0bf7917-2723-4603-ad31-2cf878faf097)

which in msgFilterRules.dat is stored as
"AND (to,contains,A) AND (subject,contains,A) OR [to,contains,B) AND (subject,contains,B] OR [to,contains,C) AND (subject,contains,C]"

That's wrong, so the filter is restored to
![image](https://github.com/Betterbird/thunderbird-patches/assets/86832723/723bdb47-1b9e-4dd5-b3d5-7ffa555ee02b)

Starting with
![image](https://github.com/Betterbird/thunderbird-patches/assets/86832723/a84f00d9-165b-4ae1-9b12-f9ff55f9154c)

is stored as
"AND (to,contains,A) AND (subject,contains,A) AND [subject,contains,dummy-dummy-dummy) OR [to,contains,B) AND (subject,contains,B] OR [to,contains,C) AND (subject,contains,C]]"

and that gets restored properly.

**Comment by Betterbird** [comment]
We took a look at our test data, since we had tested some scenarios. Turned out that we had a non-sensical test case:
![Ex4](https://github.com/Betterbird/thunderbird-patches/assets/86832723/58c4219d-8ec8-4f24-9b61-11b6306787a2)

That actually works, but as soon as you switch "all" to "any" in the inner groups, it all falls apart. So much for letting the developers do the testing :sob:

As an immediate mitigation we will disallow groups following groups, like this:
![image](https://github.com/Betterbird/thunderbird-patches/assets/86832723/b85cd8f7-f561-4225-8d9c-ab20f98c04b4)

So into a group, you need to add a term first before starting another group. That's along the lines of your suggestion (quote): _I would suggest prohibiting entering second-level nested terms at all._

Please voice objections now or it will ship on the coming Tuesday, 23rd Jan. 2024.

**Comment by Betterbird** [comment]
Yes, that will work. What doesn't work and never worked is this:
![image](https://github.com/Betterbird/thunderbird-patches/assets/86832723/75ec3655-77cb-4c5e-b42d-30cb1a23df61)

**Comment by ThiloteE** [comment]
Thank you :-)

**Comment by kupietools** [comment]
Hey there, just inquiring about the state of this. Somehow missed you flagging me back then and was just now considering giving Betterbird another show so I thought of this issue and thought I'd check up. Was it a permanent workaround, then, to disallow things like "Match all of: (Match any of A or B) AND (Match any of C or D)"?

Is that the way things currently stand? Thanks.

**Comment by Betterbird** [comment]
Hmm, we last looked at this 1.5 years ago.

The example you gave should work, see 3 and 4 comments up, but only if the first "all of" is at top level. The UI disallows what doesn't work. Please test your case.
